### ISSUE TITLE
ALERename doesn't work for CPP

### ISSUE THREAD

**Issue by bataak** [dense-analysis/ale#3914]
## Information

**VIM version**

VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Sep 19 2021 15:55:16)

Operating System: macOS version - x86_64

## What went wrong

Wanted to refactor functions with ALERename in CPP, it didn't work (without any error message). I've already selected clang-format to my fixers and it works, but it doesn't make sense ALE suggested it in info. Thanks!

### :ALEInfo

 Current Filetype: cpp
Available Linters: ['cc', 'ccls', 'clangcheck', 'clangd', 'clangtidy', 'clazy', 'cppcheck', 'cpplint', 'cquery', 'flawfinder']
   Linter Aliases:
'cc' -> ['gcc', 'clang', 'g++', 'clang++']
  Enabled Linters: ['clangd']
  Ignored Linters: []
 Suggested Fixers: 
  'astyle' - Fix C/C++ with astyle.
  'clang-format' - Fix C/C++ and cuda files with clang-format.
  'clangtidy' - Fix C/C++ and ObjectiveC files with clang-tidy.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_cpp_clangd_executable = 'clangd'
let g:ale_cpp_clangd_options = '-cross-file-rename --header-insertion=never'
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = v:null
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 1
let g:ale_completion_max_suggestions = 50
let g:ale_disable_lsp = 0
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 1
let g:ale_fixers = {'typescript': ['prettier'], 'scss': ['prettier'], 'tex': ['latexindent'], 'sh': ['shfmt'], 'html': ['prettier'], 'javascript': ['prettier'], 'cpp': ['clang-format'], 'python': ['black']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {'cpp': ['clangd'], 'tex': ['texlab'], 'javascript': ['eslint'], 'python': ['pylsp']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_root = {}
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

**Comment by hsanson** [comment]
Sorry, I do not do CPP nor use Clang tools.  From what I see in ALE it seems that installing the clangd binary and ensure is on your PATH should be enough.

**Comment by bataak** [comment]
In fact, I am currently using clangd with ycm (ycm_clangd_binary_path). Actually I think it's better to use it with ALE.

**Comment by pavel-the-best** [comment]
@hsanson I have the same issue of clangd not starting. The binary is in fact installed and it is on PATH.

**Comment by pavel-the-best** [comment]
I've isolated the problem: ALE does not launch clangd if no `compile_commands.json` file is found.
`touch compile_commands.json` fixes the problem, because clangd can now work even without `compile_commands.json`.

**Comment by hsanson** [comment]
Makes sense since clangd is a LSP and ALE requires a root directory in order to start LSPs.  From the source it seems that ALE looks for either compile_commands.json or "build", "bin" directories to determine the root directory of the project.

Appreciated if the issue is closed if it is no longer an issue.

**Comment by pavel-the-best** [comment]
It is still an issue, I don't really care what ALE wants internally, all I know is:
1) It is possible to lauch clangd in any directory with write permissions
2) It is not possible in ALE and a fix to that is undocumented and  counter intuitive. In fact, it's not even a fix but just a bypass hack.

**Comment by nospam2998** [comment]
@bataak and @p4vook,

In my attempt at going through ale's issue backlog, I see this one and agree that there are a few issues. Yet for issues to be resolved, they need to be clearly understood and a key factor to that is to describe them correctly. Could you @bataak please update the title this issue to something more appropriate for this long lived issue? Perhaps to:

> clangd does not start with: "Failed to find project root"

I believe that title would better describe what this issue really is about. Calling `:ALERename` does work. I just created this `/tmp/hello.cpp`:

```c++
#include <iostream>

int say_it() {
    std::cout << "Hello C++!" << std::endl;
}

int main() {
    say_it();
    return 0;
}
```

…with this Makefile:

```Makefile
hello: hello.cpp
	c++ -o hello hello.cpp
```

After running `bear --force-wrapper -- make` the operation in question resulted in a successful rename with the following LSP messages exchanged:

```json
{
        "method": "textDocument/rename",
        "jsonrpc": "2.0",
        "id": 2,
        "params": {
                "newName": "say_hello",
                "textDocument": {
                        "uri": "file:///tmp/c%2b%2b/hello.cpp"
                },
                "position": {
                        "character": 4,
                        "line": 2
                }
        }
}

{
        "id": 2,
        "jsonrpc": "2.0",
        "result": {
                "changes": {
                        "file:///tmp/c%2B%2B/hello.cpp": [
                                {
                                        "newText": "say_hello",
                                        "range": {
                                                "end": {
                                                        "character": 10,
                                                        "line": 2
                                                },
                                                "start": {
                                                        "character": 4,
                                                        "line": 2
                                                }
                                        }
                                },
                                {
                                        "newText": "say_hello",
                                        "range": {
                                                "end": {
                                                        "character": 10,
                                                        "line": 7
                                                },
                                                "start": {
                                                        "character": 4,
                                                        "line": 7
                                                }
                                        }
                                }
                        ]
                }
        }
}
```

As for a possible solution to the problem of no project root being found, one idea would be to set [`b:ale_root`](https://github.com/dense-analysis/ale/blob/6433d31/doc/ale.txt#L2291) to the desired directory. I believe the functionality is in place, and that the same goes for the documentation. Do you agree? It might still be too tricky to figure this stuff out though.

I'm not sure how to make the configuration option more easily discoverable. Apart from making it occur in error messages as suggested in #3481. Do you?
